File: app/src/main/java/com/example/bilalahmad/snakeandladder/MainActivity.java
package com.example.bilalahmad.snakeandladder;

import android.graphics.Point;
import android.os.Bundle;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private ImageView dice;

    private final Point SIZE = new Point();

    private boolean isOver = false;

    private int player;
    private int computer;
    private int currentPlayer;
    private final ArrayList<GameObject> snakes = new ArrayList<>();
    private final ArrayList<GameObject> ladders = new ArrayList<>();

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        dice = findViewById(R.id.dice);

        snakes.add(new GameObject(22, 3));
        snakes.add(new GameObject(14, 8));
        snakes.add(new GameObject(31, 15));
        snakes.add(new GameObject(41, 20));
        snakes.add(new GameObject(58, 37));
        snakes.add(new GameObject(67, 50));
        snakes.add(new GameObject(77, 56));
        snakes.add(new GameObject(83, 80));
        snakes.add(new GameObject(92, 76));
        snakes.add(new GameObject(99, 5));

        ladders.add(new GameObject(23, 2));
        ladders.add(new GameObject(13, 9));
        ladders.add(new GameObject(93, 17));
        ladders.add(new GameObject(54, 29));
        ladders.add(new GameObject(51, 32));
        ladders.add(new GameObject(80, 39));
        ladders.add(new GameObject(78, 62));
        ladders.add(new GameObject(44, 64));
        ladders.add(new GameObject(96, 75));
        ladders.add(new GameObject(89, 70));

        player = 1;
        computer = 1;
        currentPlayer = 1;

        getWindowManager().getDefaultDisplay().getSize(SIZE);
        findViewById(R.id.board).post(new Runnable() {
            @Override
            public void run() {
                SIZE.y = findViewById(R.id.board).getHeight();
            }
        });
    }

    public void takeTurn(View v) {

        if (!isOver) {
            int diceValue = getDice();
            if (currentPlayer == 1) {
                playersTurn(diceValue);
            }
        }
    }

    private void computerTurn(int diceValue) {
        int oldPosition = computer;
        computer += diceValue;

        if (computer == 100) {
            Toast.makeText(this, "Computer wins!", Toast.LENGTH_SHORT).show();
            isOver = true;
        }

        // Goes off the board
        else if (computer > 100) {
            computer -= diceValue;
        }

        // Snakes and ladders
        else {

            for (GameObject snake : snakes) {
                if (snake.getHead() == computer) {
                    computer = snake.getTail();
                }
            }

            for (GameObject ladder : ladders) {
                if (ladder.getTail() == computer) {
                    computer = ladder.getHead();
                }
            }
        }
        ((TextView) findViewById(R.id.computerPosition)).setText("Computer\n" + computer);
        moveOnScreen(R.id.computerPiece, oldPosition, computer);

    }

    private void playersTurn(int diceValue) {
        int oldPosition = player;
        player += diceValue;
        // Wins
        if (player == 100) {
            Toast.makeText(this, "Player wins!", Toast.LENGTH_SHORT).show();
            isOver = true;
        }

        // Goes off the board
        else if (player > 100) {
            player -= diceValue;
        }

        // Snakes and ladders
        else {

            for (GameObject snake : snakes) {
                if (snake.getHead() == player) {
                    player = snake.getTail();
                }
            }

            for (GameObject ladder : ladders) {
                if (ladder.getTail() == player) {
                    player = ladder.getHead();
                }
            }

        }
        ((TextView) findViewById(R.id.playerPosition)).setText("Player\n" + player);
        moveOnScreen(R.id.userPiece, oldPosition, player);
    }

    private int getX(int position) {
        int col = position % 10;
        col = (position % 10 == 0) ? 10 : col;

        int row = position / 10;
        row = (position % 10 == 0) ? row - 1 : row;

        col = (row % 2 != 0) ? (10 - col) : col - 1;
        return  (int) (col / 10.f * SIZE.x);
    }

    private int getY(int position) {
        int row = position / 10;
        row = (position % 10 == 0) ? row : row + 1;
        return SIZE.y - (int) (row / 10.f * SIZE.x);
    }

    private void moveOnScreen(int id, int oldPosition, int newPosition) {

        ((TextView) findViewById(R.id.gameStatus)).setText("P" + currentPlayer + " rolling dice ...");

        int x1 = getX(oldPosition);
        int y1 = getY(oldPosition);

        int x2 = getX(newPosition);
        int y2 = getY(newPosition);

        TranslateAnimation animation = new TranslateAnimation(x1, x2, y1, y2);
        animation.setDuration(500);
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {
                ((TextView) findViewById(R.id.gameStatus)).setText("Moving player " + currentPlayer + "...");
            }

            @Override
            public void onAnimationEnd(Animation animation) {
                if (currentPlayer == 1) {
                    currentPlayer = 2;
                    ((TextView) findViewById(R.id.gameStatus)).setText("Computer's Turn");

                    computerTurn(getDice());
                } else if (currentPlayer == 2) {
                    currentPlayer = 1;
                    ((TextView) findViewById(R.id.gameStatus)).setText("Player's Turn");
                }

                if (isOver) {
                    ((TextView) findViewById(R.id.gameStatus)).setText("Game Over!");
                }
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        animation.setFillAfter(true);
        findViewById(id).setVisibility(View.VISIBLE);
        findViewById(id).startAnimation(animation);
    }

    public int getDice() {

        int dice = (int) (Math.random() * 6) + 1;

        switch (dice) {

            case 1:
                this.dice.setImageResource(R.drawable.one);
                break;
            case 2:
                this.dice.setImageResource(R.drawable.two);
                break;
            case 3:
                this.dice.setImageResource(R.drawable.three);
                break;
            case 4:
                this.dice.setImageResource(R.drawable.four);
                break;
            case 5:
                this.dice.setImageResource(R.drawable.five);
                break;
            case 6:
                this.dice.setImageResource(R.drawable.six);
                break;

        }

        return dice;
    }

    //
//        public boolean onTouchEvent(MotionEvent event) {
//
//        ImageView img_animation = (ImageView) findViewById(R.id.maze);
//        int[] viewCoords = new int[2];
//        img_animation.getLocationOnScreen(viewCoords);
//
//        int touchX = (int) event.getX();
//        int touchY = (int) event.getY();
//
//        int imageX = touchX - viewCoords[0]; // viewCoords[0] is the X coordinate
//        int imageY = touchY - viewCoords[1]; // viewCoords[1] is the y coordinate
//
//        // float x = event.getX();
//        //float y = event.getY();
//        Toast.makeText(MainActivity.this,"X = "+imageX+" and y ="+imageY,Toast.LENGTH_SHORT).show();
//        return super.onTouchEvent(event);
//    }

    private class GameObject {

        private final int head;
        private final int tail;

        private GameObject(int head, int tail) {
            this.head = head;
            this.tail = tail;
        }


        public int getHead() {
            return head;
        }


        public int getTail() {
            return tail;
        }
    }
}

File: app/src/main/java/com/saifkhichi/moksha/MainActivity.kt
package com.saifkhichi.moksha

import android.graphics.Point
import android.os.Bundle
import android.view.View
import android.view.animation.Animation
import android.view.animation.Animation.AnimationListener
import android.view.animation.TranslateAnimation
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.saifkhichi.moksha.models.*
import kotlin.math.abs
import kotlin.math.roundToInt

class MainActivity : AppCompatActivity() {

    private val board = GBoard()
    private val ai = AI("Computer", R.id.computerPiece)
    private val user = User("Player", R.id.userPiece)

    private lateinit var dieUser: Die
    private lateinit var dieAI: Die
    private val SIZE = Point()
    private var isOver = false

    private var dx = 0f
    private var dy = 0f

    private var currentPlayer: GPlayer = user
        set(value) {
            field = value
            when (value) {
                is AI -> {
                    dieAI.show()
                    dieUser.hide()
                    setStatus("${value.name}'s turn.")

                    rollDice(dieAI, ai) {
                        takeTurn(ai, it)
                    }
                }
                is User -> {
                    dieAI.hide()
                    dieUser.show()
                    setStatus("Your turn. Roll the dice!")
                }
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        dieUser = Die(findViewById(R.id.dice))
        dieUser.show()

        dieAI = Die(findViewById(R.id.die_ai))
        dieAI.hide()

        currentPlayer = user
        setStatus("Your turn. Roll the dice!")

        findViewById<TextView>(R.id.computerName).text = ai.name
        findViewById<TextView>(R.id.playerName).text = user.name

        windowManager.defaultDisplay.getSize(SIZE)
        val boardContainer = findViewById<View>(R.id.board_container)
        boardContainer.post {
            dx = boardContainer.x
            dy = (0.015 * boardContainer.width).toFloat()
            SIZE.x = (0.985 * boardContainer.width).roundToInt()
            SIZE.y = (boardContainer.height * 1.6f).roundToInt()
        }
    }

    fun takeTurn(v: View?) {
        if (!isOver) {
            if (currentPlayer is User) {
                rollDice(dieUser, user) { value ->
                    takeTurn(user, value)
                }
            }
        }
    }

    private fun takeTurn(player: GPlayer, steps: Int) {
        // Move the player
        if (!player.canMove(steps)) return onMoved(0) // TODO: Move forward, then backwards
        val lastPosition = player.position
        val newPosition = lastPosition + steps

        // Run animation before moving the player
        moveBySteps(player, lastPosition, newPosition) {
            player.move(steps)

            // Check if the player has won
            if (player.hasWon()) {
                setStatus("You ${if (player is User) "won" else "lost"}! Game over.")

                // Update UI
                when (player) {
                    is AI -> findViewById<TextView>(R.id.computerPosition)
                    is User -> findViewById(R.id.playerPosition)
                }.text = "${player.position}"

                isOver = true
            } else {
                board.snakes.forEach { snake ->
                    if (player.isOn(snake)) {
                        snake.bite(player)
                    }
                }

                board.ladders.forEach { ladder ->
                    if (player.isOn(ladder)) {
                        ladder.ascend(player)
                    }
                }

                moveDirect(player, newPosition, player.position) {
                    // Update UI
                    when (player) {
                        is AI -> findViewById<TextView>(R.id.computerPosition)
                        is User -> findViewById(R.id.playerPosition)
                    }.text = "${player.position}"

                    // Check post-conditions
                    onMoved(steps)
                }
            }
        }
    }

    private fun onMoved(lastRoll: Int) {
        if (!isOver) {
            currentPlayer = when (lastRoll) {
                6 -> currentPlayer
                else -> when (currentPlayer) {
                    is User -> ai
                    is AI -> user
                }
            }
        }
    }

    private fun getX(position: Int): Int {
        var col = position % 10
        col = if (position % 10 == 0) 10 else col
        var row = position / 10
        row = if (position % 10 == 0) row - 1 else row
        col = if (row % 2 != 0) 10 - col else col - 1
        return (dx + col / 10f * SIZE.x).toInt()
    }

    private fun getY(position: Int): Int {
        var row = position / 10
        row = if (position % 10 == 0) row else row + 1
        return (SIZE.y - (row / 10f * SIZE.x) - dy).roundToInt()
    }

    private fun moveDirect(player: GPlayer, from: Int, to: Int, onMoved: (p: Int) -> Unit) = runOnUiThread {
        val x1 = getX(from)
        val y1 = getY(from)
        val x2 = getX(to)
        val y2 = getY(to)
        val animation = TranslateAnimation(x1.toFloat(), x2.toFloat(), y1.toFloat(), y2.toFloat())

        val steps = abs(to - from)
        animation.duration = when (steps == 1) {
            true -> 250L
            else -> (250 * (1 + steps / 10)).toLong()
        }
        animation.setAnimationListener(object : AnimationListener {
            override fun onAnimationStart(animation: Animation) {
                setStatus("Moving ${if (player is User) "your" else "computer's"} piece...")
            }

            override fun onAnimationEnd(animation: Animation) {
                onMoved(to)
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        animation.fillAfter = true
        findViewById<View>(player.view).visibility = View.VISIBLE
        findViewById<View>(player.view).startAnimation(animation)
    }

    private fun moveBySteps(player: GPlayer, current: Int, final: Int, onEnd: () -> Unit) {
        var next = if (final > current) current + 1 else current - 1
        moveDirect(player, current, next) { position ->
            if (position != final) {
                next = if (final > current) current + 1 else current - 1
                moveBySteps(player, position, final, onEnd)
            } else {
                onEnd()
            }
        }
    }

    private fun rollDice(die: Die, player: GPlayer, callback: (Int) -> Unit) {
        setStatus("${player.name} rolling the dice...")
        die.roll { value ->
            setStatus("${player.name} rolled ${value}...")
            callback(value)
        }
    }

    private fun setStatus(status: String) = runOnUiThread {
        findViewById<TextView>(R.id.gameStatus).text = ">> $status"
    }

}